[Reliability] Introduce local vs store timezone difference tracks

ThomazFB opened the pull request:
Closes: #9835

Why
==========
We want to verify how often our user base uses a different device timezone from the one configured in the store. To achieve this a new track event called `DASHBOARD_STORE_TIMEZONE_DIFFER_FROM_DEVICE` must be introduced. The relevance of such data is that since the app can adopt the local device timezone to fetch stats, whenever the store is configured with a different timezone, the analytics values and numbers between the wp-admin and app stats can significantly differ.

How
==========
 This PR introduces a track event triggered whenever we start the app Dashboard and the device's local timezone configuration differs from the store timezone configuration. The GMT offset will be compared in seconds, but the tracks will log it in hours for better analytics data navigation.


How to Test
==========
### Scenario 1
1. Open the app with a store configured in a different timezone from the device you're testing
2. Verify that the `DASHBOARD_STORE_TIMEZONE_DIFFER_FROM_DEVICE` event is triggered as soon as the `My Store` section is loaded
3. Verify that the same event contains the `store_timezone` and `local_timezone` carrying the expected values from each
4. Now, without reopening the app, go to the App More menu and switch the store to another one with a different timezone from the local
5. Verify that again the `DASHBOARD_STORE_TIMEZONE_DIFFER_FROM_DEVICE` event is triggered, containing the same `local_timezone` as before, and that `store_timezone` follows the value of the new selected store

### Scenario 2
1. Open the app with a store configured with the same timezone from the device you're testing
2. Verify that the `DASHBOARD_STORE_TIMEZONE_DIFFER_FROM_DEVICE` is NOT triggered


Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Define basic stats tracking trigger function inside DashboardViewModel

ThomazFB committed:
Define dashboardStoreTimezoneDifferFromDevice track event

ThomazFB committed:
Set WooAnalyticsEvent function for timezone tracks event creation

ThomazFB committed:
Define missing keys for new Dashboard tracks event

ThomazFB committed:
Configure WooAnalyticEvent creation from expected function

ThomazFB committed:
Trigger expected analytics trigger when condition is met

ThomazFB committed:
Trigger timezone tracks event when Stats view loads

ThomazFB committed:
Refactor parameter and timezone acquisition logic

ThomazFB committed:
Adjust timezone track to send GMT offset as hours instead of seconds

ThomazFB committed:
Introduce basic test for timezone event trigger

ThomazFB committed:
Complete timezone unit test scenarios

rachelmcr commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift:
```diff
@@ -180,6 +180,18 @@ final class DashboardViewModel {
+        }
+
+        let localGMTOffsetInHours = String(localTimezone.secondsFromGMT() / 3600)
+        let siteGMTOffsetInHours = String(siteTimezone.secondsFromGMT() / 3600)
```
I'd suggest converting the `secondsFromGMT` offsets to `Double` values for this calculation. Otherwise, these are calculated as `Int` and any timezones that aren't a full hour offset are rounded (e.g. `Asia/Tehran` is +3.5 hours from GMT but the offset calculated here for that timezone is `3`).

(Sidenote: We also have the site's GMT offset in `ServiceLocator.stores.sessionManager.defaultSite?.gmtOffset`, if we wanted to use that rather than calculating it again here.)

ThomazFB replied:
Very good point! I forgot to account for the non-integer timezones, thanks! 🙇 I fixed it in https://github.com/woocommerce/woocommerce-ios/pull/9848/commits/65d8f72186891328ebd2a03d2558e1913b9e2112 and https://github.com/woocommerce/woocommerce-ios/pull/9848/commits/2464c35197fb599a5debd75f0c2070bf4a956716. And added a specific unit test for this scenario in https://github.com/woocommerce/woocommerce-ios/pull/9848/commits/9612817819a7150ca8df0582fd42d8d3af76542b

rachelmcr commented on WooCommerce/Classes/Analytics/WooAnalyticsEvent+Dashboard.swift:
```diff
@@ -30,6 +33,15 @@ extension WooAnalyticsEvent {
+
+        /// Tracked when the dashboard is accessed with a device timezone different from the store timezone.
+        /// - Parameter localTimezone: The current timezone of the device running the app.
+        /// - Parameter storeTimezone: The store timezone defined by the API.
```
Maybe we should document here that we're not recording the timezone identifier but rather the timezone offset from GMT. Perhaps parameter names like `localTimezoneOffset` and `storeTimezoneOffset` would make that clearer too?

ThomazFB replied:
That makes sense, thanks! Updated in https://github.com/woocommerce/woocommerce-ios/pull/9848/commits/001368f53a0257305ab8616a5d52cecd5adf6555

rachelmcr reviewed (commented):
Nice idea to track the timezone difference like this! Is the Tracks event already validated and registered?

I have a concern about the timing, though: I was testing while logged in via the REST API (with a store not using Jetpack). When `trackStatsTimezone(localTimezone:siteTimezone:)` was called, `ServiceLocator.stores.sessionManager.defaultSite` was still `nil` and so `TimeZone.siteTimezone` defaulted to `.current` instead of using the site's actual timezone. Because of that, it didn't trigger the Tracks event even though the site timezone was different from the local timezone.

When `DefaultStoreManager` is initialized, it fetches the site info and sets `defaultSite` with the action `WordPressSiteAction.fetchSiteInfo` in `DefaultStoreManager.restoreSessionSite(with:)`. However, that action isn't necessarily completed in time to use it with this timing.

Instead of triggering the Tracks event on `StoreStatsAndTopPerformersViewController.viewDidLoad()`, could we delay it to a point where we're more certain that `defaultSite` is set?

ThomazFB committed:
Adjust offset to be calculated as double

ThomazFB committed:
Format decimals from offset value

ThomazFB committed:
Add decimal timezone test scenario

ThomazFB committed:
Update timezone track event documentation

ThomazFB committed:
Move timezone track event trigger to site observation inside the DashboardVIewController

ThomazFB committed:
Take advantage of available site gmtOffset field during track event

ThomazFB committed:
Update timezone track event tests inside DashboardViewModelTests

ThomazFB commented:
Thank you @rachelmcr, for your great insights and review 🙇. Once again, your input pushed me to find better solutions 😄. I addressed all your suggestions, and I believe the PR is ready for a second round! 

>Instead of triggering the Tracks event on StoreStatsAndTopPerformersViewController.viewDidLoad(), could we delay it to a point where we're more certain that defaultSite is set?

Thank you for finding this corner case, your suggestion made me investigate a little bit when we track this event, and I found that WCiOS also have a similar reactive solution to support what I did for this track on the [Android implementation](https://github.com/woocommerce/woocommerce-android/pull/9139). I decided to move the triggering for the published site field observation, and now we can trigger the track event when we're sure we have the site data. I introduced the change in https://github.com/woocommerce/woocommerce-ios/pull/9848/commits/1a4621ea2d769dfb1146a558e24f04e44b3fe3f1.

>We also have the site's GMT offset in ServiceLocator.stores.sessionManager.defaultSite?.gmtOffset, if we wanted to use that rather than calculating it again here

Following the solution I mentioned before, I implemented your suggestion, but instead of using the ServiceLocator, I accessed the `gmtOffset` directly from the provided site observation result. WDYT? I introduced the change in https://github.com/woocommerce/woocommerce-ios/pull/9848/commits/4f536f5fb8d45cb07980e348701e32887f258b5a

ThomazFB commented:
>Is the Tracks event already validated and registered?

I already validated and registered the Android one, but the iOS event didn't show up in the tracks dashboard yet, so I'm waiting to move forward with it, I hope it's ok! 😬

rachelmcr reviewed (approved):
These changes look great! The new timing and approach makes a lot of sense, and tests well. 🙌 

> the iOS event didn't show up in the tracks dashboard yet, so I'm waiting to move forward with it

Sure thing! As long as that's on your radar 👍 You can also use the Tracks Live view for this, since it will show up there first (earlier than it's added to the Events list). I tend to check that Live view anyway just to be sure that new events aren't being rejected.

ThomazFB commented:
>Sure thing! As long as that's on your radar 👍 You can also use the Tracks Live view for this, since it will show up there first (earlier than it's added to the Events list). I tend to check that Live view anyway just to be sure that new events aren't being rejected.

Oh, good to know! I'll trigger some of those events today and catch them on the Tracks Live. Thank you! 😄
